extensions/interactions/CodeRepl/CodeRepl.js

```diff
@@ -146,3 +146,156 @@ oppia.directive('oppiaShortResponseCodeRepl', [
+          if (key > numSpaces) {
+            delete numSpaceToDesiredIndentLevel[key];
+          }
+        }
+
+        if (isShortfallLine) {
+          numSpaces = Object.keys(numSpaceToDesiredIndentLevel).reduce(
+            function(max, key) {
+              return Math.max(numSpaceToDesiredIndentLevel[key], max);
+            }, -Infinity);
+        }
+
+        var normalizedLine = '';
+        for (var i = 0; i < numSpaceToDesiredIndentLevel[numSpaces]; i++) {
+          normalizedLine += FOUR_SPACES;
+        }
+        normalizedLine += removeLeadingWhitespace(line);
+        normalizedCodeLines.push(normalizedLine);
+      });
+      return normalizedCodeLines.join('\n');
+    },
+    // A very naive approach to 'normalizing' the code is to strip out all
+    // comments and whitespace. This normalization currently assumes Python.
+    normalizePythonCode: function(code) {
+      // TODO(sll): This does not correctly handle the case where '#' is
+      // within quotes, or where it is escaped.
+      // remove comments
+      var strippedCode = code.replace(/#.*/g, '');
+      // remove whitespace
+      return strippedCode.replace(/\s+/g, '');
+    }
+  };
+}]);
+
+oppia.factory('codeReplRulesService', [
+    '$filter', 'codeReplNormalizationService',
+    function($filter, codeReplNormalizationService) {
+  return {
+    CodeEquals: function(answer, inputs) {
+      var normalizedCode =
+        codeReplNormalizationService.getNormalizedCode(answer.code);
+      var normalizedExpectedCode =
+        codeReplNormalizationService.getNormalizedCode(inputs.x);
+      return normalizedCode == normalizedExpectedCode;
+    },
+    CodeContains: function(answer, inputs) {
+      var normalizedCode =
+        codeReplNormalizationService.getNormalizedCode(answer.code);
+      var normalizedSnippet =
+        codeReplNormalizationService.getNormalizedCode(inputs.x);
+      return normalizedCode.indexOf(normalizedSnippet) != -1;
+    },
+    CodeDoesNotContain: function(answer, inputs) {
+      var normalizedCode =
+        codeReplNormalizationService.getNormalizedCode(answer.code);
+      var normalizedSnippet =
+        codeReplNormalizationService.getNormalizedCode(inputs.x);
+      return normalizedCode.indexOf(normalizedSnippet) == -1;
+    },
+    OutputEquals: function(answer, inputs) {
+      var normalizedOutput = $filter('normalizeWhitespace')(answer.output);
+      var normalizedExpectedOutput =
+        $filter('normalizeWhitespace')(inputs.x);
+      return normalizedOutput == normalizedExpectedOutput;
+    },
+    ResultsInError: function(answer, inputs) {
+      return !!(answer.error.trim());
+    },
+    ErrorContains: function(answer, inputs) {
+      var normalizedError = $filter('normalizeWhitespace')(answer.error);
+      var normalizedSnippet = $filter('normalizeWhitespace')(inputs.x);
+      return normalizedError.indexOf(normalizedSnippet) != -1;
+    },
+    FuzzyMatches: function(answer, inputs) {
```

Comment:
Oh wait, sorry, scrap that. I realize we did do it for InteractiveMap.

I am concerned about this, though -- it's a bit half-baked. Can we remove it completely for now? /cc @BenHenning